diff -up Linux-PAM-1.1.1/modules/pam_unix/pam_unix_passwd.c.remember Linux-PAM-1.1.1/modules/pam_unix/pam_unix_passwd.c
--- Linux-PAM-1.1.1/modules/pam_unix/pam_unix_passwd.c.remember	2012-07-19 14:37:25.000000000 +0200
+++ Linux-PAM-1.1.1/modules/pam_unix/pam_unix_passwd.c	2012-10-12 09:25:30.708991111 +0200
@@ -279,13 +279,15 @@ static int check_old_password(const char
 	char *s_luser, *s_uid, *s_npas, *s_pas;
 	int retval = PAM_SUCCESS;
 	FILE *opwfile;
+	size_t len = strlen(forwho);
 
 	opwfile = fopen(OLD_PASSWORDS_FILE, "r");
 	if (opwfile == NULL)
 		return PAM_ABORT;
 
 	while (fgets(buf, 16380, opwfile)) {
-		if (!strncmp(buf, forwho, strlen(forwho))) {
+		if (!strncmp(buf, forwho, len) && (buf[len] == ':' ||
+			buf[len] == ',')) {
 			char *sptr;
 			buf[strlen(buf) - 1] = '\0';
 			s_luser = strtok_r(buf, ":,", &sptr);
diff -up Linux-PAM-1.1.1/modules/pam_unix/passverify.c.remember Linux-PAM-1.1.1/modules/pam_unix/passverify.c
--- Linux-PAM-1.1.1/modules/pam_unix/passverify.c.remember	2009-12-08 10:20:27.000000000 +0100
+++ Linux-PAM-1.1.1/modules/pam_unix/passverify.c	2012-10-12 09:24:54.249180106 +0200
@@ -562,6 +562,7 @@ save_old_password(pam_handle_t *pamh, co
     int found = 0;
     struct passwd *pwd = NULL;
     struct stat st;
+    size_t len = strlen(forwho);
 #ifdef WITH_SELINUX
     security_context_t prev_context=NULL;
 #endif
@@ -629,7 +630,7 @@ save_old_password(pam_handle_t *pamh, co
     }
 
     while (fgets(buf, 16380, opwfile)) {
-	if (!strncmp(buf, forwho, strlen(forwho))) {
+	if (!strncmp(buf, forwho, len) && strchr(":,\n", buf[len]) != NULL) {
 	    char *sptr = NULL;
 	    found = 1;
 	    if (howmany == 0)
