diff -up Linux-PAM-1.1.1/modules/pam_selinux/pam_selinux.c.canonicalize Linux-PAM-1.1.1/modules/pam_selinux/pam_selinux.c
--- Linux-PAM-1.1.1/modules/pam_selinux/pam_selinux.c.canonicalize	2014-06-19 13:01:55.000000000 +0200
+++ Linux-PAM-1.1.1/modules/pam_selinux/pam_selinux.c	2014-06-20 14:30:04.244733946 +0200
@@ -595,6 +595,7 @@ pam_sm_open_session(pam_handle_t *pamh,
   const void *void_service;
   const char *service;
 #endif
+  const struct passwd *pwd;
 
   /* Parse arguments. */
   for (i = 0; i < argc; i++) {
@@ -642,6 +643,11 @@ pam_sm_open_session(pam_handle_t *pamh,
   }
   username = void_username;
 
+  if ((pwd = pam_modutil_getpwnam(pamh, username)) != NULL) {
+    username = pwd->pw_name;
+  } /* ignore error and keep using original username */
+
+
 #ifdef HAVE_GETSEUSER
   if (pam_get_item(pamh, PAM_SERVICE, (void *) &void_service) != PAM_SUCCESS ||
                    void_service == NULL) {
