diff -up Linux-PAM-1.1.1/modules/pam_access/pam_access.c.netgroup Linux-PAM-1.1.1/modules/pam_access/pam_access.c
--- Linux-PAM-1.1.1/modules/pam_access/pam_access.c.netgroup	2014-06-19 14:59:47.000000000 +0200
+++ Linux-PAM-1.1.1/modules/pam_access/pam_access.c	2014-06-20 15:21:14.300815760 +0200
@@ -471,20 +471,18 @@ netgroup_match (pam_handle_t *pamh, cons
   int retval;
   char *mydomain = NULL;
 
-#ifdef HAVE_YP_GET_DEFAUTL_DOMAIN
-  yp_get_default_domain(&mydomain);
-#elif defined(HAVE_GETDOMAINNAME)
+#if defined(HAVE_GETDOMAINNAME)
   char domainname_res[256];
 
   if (getdomainname (domainname_res, sizeof (domainname_res)) == 0)
     {
-      if (strcmp (domainname_res, "(none)") == 0)
+      if (domainname_res[0] != '\0' && strcmp (domainname_res, "(none)") != 0)
         {
-          /* If domainname is not set, some systems will return "(none)" */
-	  domainname_res[0] = '\0';
-	}
-      mydomain = domainname_res;
+          mydomain = domainname_res;
+        }
     }
+#elif defined(HAVE_YP_GET_DEFAULT_DOMAIN)
+  yp_get_default_domain(&mydomain);
 #endif
 
 #ifdef HAVE_INNETGR
@@ -523,7 +521,11 @@ user_match (pam_handle_t *pamh, char *to
      * name of the user's primary group.
      */
 
-    if ((at = strchr(tok + 1, '@')) != 0) {	/* split user@host pattern */
+    /* Try to split on a pattern (@*[^@]+)(@+.*) */
+    for (at = tok; *at == '@'; ++at);
+
+    if ((at = strchr(at, '@')) != NULL) {
+        /* split user@host pattern */
 	if (item->hostname == NULL)
 	    return NO;
 	memcpy (&fake_item, item, sizeof(fake_item));
