diff -up Linux-PAM-1.1.1/modules/pam_pwhistory/opasswd.c.opasswd-tolerant Linux-PAM-1.1.1/modules/pam_pwhistory/opasswd.c
--- Linux-PAM-1.1.1/modules/pam_pwhistory/opasswd.c.opasswd-tolerant	2014-06-20 15:52:33.445936189 +0200
+++ Linux-PAM-1.1.1/modules/pam_pwhistory/opasswd.c	2014-07-17 17:23:06.765713027 +0200
@@ -82,10 +82,15 @@ parse_entry (char *line, opwd *data)
 {
   const char delimiters[] = ":";
   char *endptr;
+  char *count;
 
   data->user = strsep (&line, delimiters);
   data->uid = strsep (&line, delimiters);
-  data->count = strtol (strsep (&line, delimiters), &endptr, 10);
+  count = strsep (&line, delimiters);
+  if (data->user == NULL || data->uid == NULL || count == NULL)
+      return 1;
+
+  data->count = strtol (count, &endptr, 10);
   if (endptr != NULL && *endptr != '\0')
       return 1;
 
diff -up Linux-PAM-1.1.1/modules/pam_unix/passverify.c.opasswd-tolerant Linux-PAM-1.1.1/modules/pam_unix/passverify.c
--- Linux-PAM-1.1.1/modules/pam_unix/passverify.c.opasswd-tolerant	2014-06-20 15:52:33.447936235 +0200
+++ Linux-PAM-1.1.1/modules/pam_unix/passverify.c	2014-07-17 17:23:06.767713076 +0200
@@ -639,11 +639,23 @@ save_old_password(pam_handle_t *pamh, co
 	    	continue;
 	    buf[strlen(buf) - 1] = '\0';
 	    s_luser = strtok_r(buf, ":", &sptr);
+	    if (s_luser == NULL) {
+		found = 0;
+		continue;
+	    }
 	    s_uid = strtok_r(NULL, ":", &sptr);
+	    if (s_uid == NULL) {
+		found = 0;
+		continue;
+	    }
 	    s_npas = strtok_r(NULL, ":", &sptr);
+	    if (s_npas == NULL) {
+		found = 0;
+		continue;
+	    }
 	    s_pas = strtok_r(NULL, ":", &sptr);
 	    npas = strtol(s_npas, NULL, 10) + 1;
-	    while (npas > howmany) {
+	    while (npas > howmany && s_pas != NULL) {
 		s_pas = strpbrk(s_pas, ",");
 		if (s_pas != NULL)
 		    s_pas++;
