diff -up Linux-PAM-1.1.1/modules/pam_access/pam_access.c.group-first Linux-PAM-1.1.1/modules/pam_access/pam_access.c
--- Linux-PAM-1.1.1/modules/pam_access/pam_access.c.group-first	2016-01-07 17:31:33.330924033 +0100
+++ Linux-PAM-1.1.1/modules/pam_access/pam_access.c	2016-11-01 16:05:29.270129502 +0100
@@ -524,7 +524,9 @@ user_match (pam_handle_t *pamh, char *to
     /* Try to split on a pattern (@*[^@]+)(@+.*) */
     for (at = tok; *at == '@'; ++at);
 
-    if ((at = strchr(at, '@')) != NULL) {
+    if (tok[0] == '(' && tok[strlen(tok) - 1] == ')') {
+      return (group_match (pamh, tok, string, item->debug));
+    } else if ((at = strchr(at, '@')) != NULL) {
         /* split user@host pattern */
 	if (item->hostname == NULL)
 	    return NO;
@@ -549,9 +551,7 @@ user_match (pam_handle_t *pamh, char *to
 		hostname = item->hostname;
 	}
         return (netgroup_match (pamh, tok + 1, hostname, string, item->debug));
-    } else if (tok[0] == '(' && tok[strlen(tok) - 1] == ')')
-      return (group_match (pamh, tok, string, item->debug));
-    else if ((rv=string_match (pamh, tok, string, item->debug)) != NO) /* ALL or exact match */
+    } else if ((rv=string_match (pamh, tok, string, item->debug)) != NO) /* ALL or exact match */
       return rv;
     else if (item->only_new_group_syntax == NO &&
 	     pam_modutil_user_in_group_nam_nam (pamh,
