diff -up Linux-PAM-1.1.1/modules/pam_limits/limits.conf.nofile Linux-PAM-1.1.1/modules/pam_limits/limits.conf
--- Linux-PAM-1.1.1/modules/pam_limits/limits.conf.nofile	2008-01-08 15:49:05.000000000 +0100
+++ Linux-PAM-1.1.1/modules/pam_limits/limits.conf	2014-06-19 15:35:10.460863266 +0200
@@ -6,7 +6,7 @@
 #
 #Where:
 #<domain> can be:
-#        - an user name
+#        - a user name
 #        - a group name, with @group syntax
 #        - the wildcard *, for default entry
 #        - the wildcard %, can be also used with %group syntax,
@@ -21,7 +21,7 @@
 #        - data - max data size (KB)
 #        - fsize - maximum filesize (KB)
 #        - memlock - max locked-in-memory address space (KB)
-#        - nofile - max number of open files
+#        - nofile - max number of open file descriptors
 #        - rss - max resident set size (KB)
 #        - stack - max stack size (KB)
 #        - cpu - max CPU time (MIN)
diff -up Linux-PAM-1.1.1/modules/pam_limits/limits.conf.5.xml.nofile Linux-PAM-1.1.1/modules/pam_limits/limits.conf.5.xml
--- Linux-PAM-1.1.1/modules/pam_limits/limits.conf.5.xml.nofile	2014-06-19 14:41:38.000000000 +0200
+++ Linux-PAM-1.1.1/modules/pam_limits/limits.conf.5.xml	2014-06-19 15:34:40.222153292 +0200
@@ -178,7 +178,7 @@
             <varlistentry>
               <term><option>nofile</option></term>
               <listitem>
-                <para>maximum number of open files</para>
+                <para>maximum number of open file descriptors</para>
               </listitem>
             </varlistentry>
             <varlistentry>
