diff -up Linux-PAM-1.1.1/modules/pam_tally2/pam_tally2.c.executable Linux-PAM-1.1.1/modules/pam_tally2/pam_tally2.c
--- Linux-PAM-1.1.1/modules/pam_tally2/pam_tally2.c.executable	2016-01-07 17:23:16.000000000 +0100
+++ Linux-PAM-1.1.1/modules/pam_tally2/pam_tally2.c	2016-01-07 17:29:58.780797959 +0100
@@ -384,7 +384,7 @@ get_tally(pam_handle_t *pamh, uid_t uid,
 
     lstat_ret = lstat(filename, &fileinfo);
     if (lstat_ret) {
-      *tfile=open(filename, O_APPEND|O_CREAT, 0700);
+      *tfile=open(filename, O_APPEND|O_CREAT, S_IRUSR|S_IWUSR);
       /* Create file, or append-open in pathological case. */
       if (*tfile == -1) {
 #ifndef MAIN
