diff -up Linux-PAM-1.1.1/modules/pam_userdb/pam_userdb.c.crypt-hash Linux-PAM-1.1.1/modules/pam_userdb/pam_userdb.c
--- Linux-PAM-1.1.1/modules/pam_userdb/pam_userdb.c.crypt-hash	2008-11-30 18:13:58.000000000 +0100
+++ Linux-PAM-1.1.1/modules/pam_userdb/pam_userdb.c	2014-07-14 15:36:00.038609996 +0200
@@ -213,24 +213,23 @@ user_lookup (pam_handle_t *pamh, const c
 	  /* crypt(3) password storage */
 
 	  char *cryptpw;
-	  char salt[2];
 
-	  if (data.dsize != 13) {
+	  if (data.dsize < 13) {
 	    compare = -2;
 	  } else if (ctrl & PAM_ICASE_ARG) {
 	    compare = -2;
 	  } else {
-	    salt[0] = *data.dptr;
-	    salt[1] = *(data.dptr + 1);
+	    cryptpw = crypt (pass, data.dptr);
 
-	    cryptpw = crypt (pass, salt);
-
-	    if (cryptpw) {
-	      compare = strncasecmp (data.dptr, cryptpw, data.dsize);
+	    if (cryptpw && strlen(cryptpw) == (size_t)data.dsize) {
+	      compare = memcmp(data.dptr, cryptpw, data.dsize);
 	    } else {
 	      compare = -2;
 	      if (ctrl & PAM_DEBUG_ARG) {
-		pam_syslog(pamh, LOG_INFO, "crypt() returned NULL");
+		if (cryptpw)
+		  pam_syslog(pamh, LOG_INFO, "lengths of computed and stored hashes differ");
+		else
+		  pam_syslog(pamh, LOG_INFO, "crypt() returned NULL");
 	      }
 	    };
 
